cmd/pomerium-cli/cache.go

```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+func configHome() string {
+	cfgDir, err := os.UserConfigDir()
+	if err != nil {
+		fatalf("error getting user config dir: %v", err)
+	}
+
+	ch := filepath.Join(cfgDir, "pomerium-cli")
+	err = os.MkdirAll(ch, 0755)
+	if err != nil {
+		fatalf("error creating user config dir: %v", err)
+	}
+
+	return ch
+}
+
+func cachePath() string {
+	return filepath.Join(configHome(), "cache", "exec-credential")
+}
+
+func cachedCredentialPath(serverURL string) string {
+	h := sha256.New()
+	_, _ = h.Write([]byte(serverURL))
+	id := hex.EncodeToString(h.Sum(nil))
+	return filepath.Join(cachePath(), id+".json")
+}
+
+func loadCachedCredential(serverURL string) *ExecCredential {
+	fn := cachedCredentialPath(serverURL)
+
+	f, err := os.Open(fn)
+	if err != nil {
+		return nil
+	}
+	defer f.Close()
+
+	var creds ExecCredential
+	err = json.NewDecoder(f).Decode(&creds)
+	if err != nil {
+		_ = os.Remove(fn)
+		return nil
+	}
+
+	if creds.Status == nil {
+		_ = os.Remove(fn)
+		return nil
+	}
+
+	ts := creds.Status.ExpirationTimestamp
+	if ts.IsZero() || ts.Before(time.Now()) {
+		_ = os.Remove(fn)
+		return nil
+	}
+
+	return &creds
+}
+
+func saveCachedCredential(serverURL string, creds *ExecCredential) {
+	fn := cachedCredentialPath(serverURL)
+
+	_ = os.MkdirAll(filepath.Dir(fn), 0755)
+
+	f, err := os.Create(fn)
+	if err != nil {
+		return
+	}
+	defer f.Close()
```

Comment:
Since when you write to `f`, you must check the error when closing it, to make sure data was written successfully.

Reply:
We are ignoring all errors in this function. What do you want me to do if there is an error.

Reply:
@calebdoxsey I think just emit a warning the saving credentials failed. So user won't be surprise.

Reply:
We will now log errors.

FWIW as far as I know, writing files is a synchronous operation in Go. Any write errors will be surfaced in the call to `.Write`.

Reply:
@calebdoxsey Yes, but what I mean is if `f.Close()` error, then everything you wrote previously will be lost. That's why you always have to check for error return by `f.Close()` _if_ you wrote something to file. Reading from file does not have this restriction.